nifi-nar-bundles/nifi-pulsar-bundle/nifi-pulsar-processors/src/main/java/org/apache/nifi/processors/pulsar/AbstractPulsarProducerProcessor.java

```diff
@@ -0,0 +1,323 @@
+
+    static {
+        final List<PropertyDescriptor> properties = new ArrayList<>();
+        properties.add(PULSAR_CLIENT_SERVICE);
+        properties.add(TOPIC);
+        properties.add(ASYNC_ENABLED);
+        properties.add(MAX_ASYNC_REQUESTS);
+        properties.add(BATCHING_ENABLED);
+        properties.add(BATCHING_MAX_MESSAGES);
+        properties.add(BATCH_INTERVAL);
+        properties.add(BLOCK_IF_QUEUE_FULL);
+        properties.add(COMPRESSION_TYPE);
+        properties.add(MESSAGE_ROUTING_MODE);
+        properties.add(PENDING_MAX_MESSAGES);
+        PROPERTIES = Collections.unmodifiableList(properties);
+
+        final Set<Relationship> relationships = new HashSet<>();
+        relationships.add(REL_SUCCESS);
+        relationships.add(REL_FAILURE);
+        RELATIONSHIPS = Collections.unmodifiableSet(relationships);
+    }
+
+    @Override
+    public Set<Relationship> getRelationships() {
+        return RELATIONSHIPS;
+    }
+
+    @Override
+    protected List<PropertyDescriptor> getSupportedPropertyDescriptors() {
+        return PROPERTIES;
+    }
+
+    protected LRUCache<String, Producer<T>> producers;
+
+    // Pool for running multiple publish Async requests
+    protected ExecutorService publisherPool;
+    protected ExecutorCompletionService<Object> publisherService;
+
+    @OnScheduled
+    public void init(ProcessContext context) {
+        if (context.getProperty(ASYNC_ENABLED).isSet() && context.getProperty(ASYNC_ENABLED).asBoolean()) {
+            publisherPool = Executors.newFixedThreadPool(context.getProperty(MAX_ASYNC_REQUESTS).asInteger());
+            publisherService = new ExecutorCompletionService<>(publisherPool);
+        }
+    }
+
+    @OnUnscheduled
+    public void shutDown(final ProcessContext context) {
+       if (context.getProperty(ASYNC_ENABLED).isSet() && context.getProperty(ASYNC_ENABLED).asBoolean()) {
+           // Stop all the async publishers
+           try {
+              publisherPool.shutdown();
+              publisherPool.awaitTermination(20, TimeUnit.SECONDS);
+           } catch (InterruptedException e) {
+              getLogger().error("Unable to stop all the Pulsar Producers", e);
+           }
+       }
+    }
+
+    @OnStopped
+    public void cleanUp(final ProcessContext context) {
+       getProducerCache(context).clear();
+    }
+
+    @SuppressWarnings("rawtypes")
+    protected void sendAsync(Producer producer, final ProcessSession session, final byte[] messageContent) {
+        try {
+            publisherService.submit(new Callable<Object>() {
+               @Override
+               public Object call() throws Exception {
+                 try {
+                     return producer.newMessage().value(messageContent).sendAsync().handle((msgId, ex) -> {
+                        if (msgId != null) {
+                           return msgId;
+                        } else {
+                           FlowFile flowFile = session.create();
```

Comment:
When I tried to use the original FlowFile, I would get runtime exceptions about the FlowFile NOT being successfully transferred to a relationship. Since this code is inside an async callback method, by the time it is invoked the original session has been closed. Therefore, my only option to transfer out the failed messages is to create a new FlowFile specifically for failures.

Reply:
You'll need to avoid committing the session until you've received an ACK/NACK for each message in that session. As-is, this is not transferring out the failed messages. What it is doing is creating a new, empty FlowFile that is not associated with the incoming FlowFile, and sending that empty FlowFile to failure.

Reply:
I am not sure to accomplish this, since after I submit the callback method to the ExecutorService, the original calling thread returns to the AbstractProcessor base class' `onTrigger()` method, which calls makes the call to `session.commit();`

Reply:
You'll have two options - you can either extend AbstractSessionFactoryProcessor, instead of AbstractProcessor, and manage the sessions yourself (in which case you just have to ensure that in all cases the session is properly rolled back or committed), or you could wait on a `Future`, perhaps, to avoid returning from `onTrigger` until the messages have been acknowledged.

Reply:
I have updated the code to use a countdown latch for all async message handling, this ensures I have confirmation of message hand-off to the broker for ALL messages in the session, prior to routing the session to SUCCESS or FAILURE. 

I also added message demarcation, as async processing of a single message doesn't make much sense, as I am forced to wait for a response from the broker before committing the session anyway. Therefore, I decided that a useful scenario would be to have async processing of a demarcated message, with a countdown latch to ensure all the messages are received before proceeding.

Reply:
I am, seeing the following error in the "continuous-integration/appveyor/pr — AppVeyor " output:

[ERROR] The build could not read 1 project -> [Help 1]
1066[ERROR]   
1067[ERROR]   The project org.apache.nifi:nifi-pulsar-bundle:1.7.0-SNAPSHOT (C:\projects\nifi\nifi-nar-bundles\nifi-pulsar-bundle\pom.xml) has 1 error
1068[ERROR]     Non-resolvable parent POM: Could not find artifact org.apache.nifi:nifi-nar-bundles:pom:1.7.0-SNAPSHOT and 'parent.relativePath' points at wrong local POM @ line 19, column 13 -> [Help 2]

Does this mean the parent POM changed?  Is there anything else you need from me to move this PR forward?